File: packages/open-protocol/openProtocol.js
/**
 * Open Protocol Helper
 *
 * Centralizes frame construction, parsing, and MID definitions
 * for Atlas Copco Open Protocol communication.
 *
 * Frame format (20-byte header):
 *   Bytes 0-3   : Length (4 digits, includes header + body + NUL)
 *   Bytes 4-7   : MID (4 digits)
 *   Bytes 8-10  : Revision (3 digits)
 *   Bytes 11-12 : NoAck flag (2 digits, unused here)
 *   Bytes 13-14 : Station ID (2 digits, unused here)
 *   Bytes 15-16 : Spindle ID (2 digits, unused here)
 *   Bytes 17-19 : Spare (3 digits)
 *   Byte  20    : NUL terminator
 */

const HEADER_LENGTH = 20;
const NUL = "\x00";

// --- Frame Builder ---

/**
 * Build a raw Open Protocol frame string.
 * @param {string} mid - 4-digit MID (e.g. "0001")
 * @param {string} [revision="001"] - 3-digit revision
 * @param {string} [body=""] - Optional body after header
 * @returns {Buffer}
 */
function buildFrame(mid, revision = "001", body = "") {
  const totalLength = HEADER_LENGTH + body.length + 1; // +1 for NUL
  const lengthStr = String(totalLength).padStart(4, "0");
  const revStr = String(revision).padStart(3, "0");
  const header = `${lengthStr}${mid}${revStr}${" ".repeat(9)}`;
  return Buffer.from(header + body + NUL, "ascii");
}

// --- MID Builders ---

/** MID 0001 - Communication Start */
function buildMID0001() {
  return Buffer.from("00200001001         \x00", "ascii");
}

/** MID 0050 - Vehicle ID Number Download Request */
function buildMID0050(vin) {
  const vinStr = vin.padEnd(25, " ").substring(0, 25);
  const frameStr = `00450050            ${vinStr}\x00`;
  return Buffer.from(frameStr, "ascii");
}

/** MID 0060 - Last Tightening Result Data Subscribe */
function buildMID0060() {
  // Revision 001, extra spacing for subscribe frame
  return Buffer.from(`002000600011        ${NUL}`, "ascii");
}

/** MID 0062 - Last Tightening Result Data Acknowledge */
function buildMID0062() {
  return Buffer.from("00200062            \x00", "ascii");
}

/** MID 9999 - Alive / Keep-Alive (Heartbeat) */
function buildMID9999() {
  return Buffer.from("00209999001         \x00", "ascii");
}

// --- Parser ---

/**
 * Parse an Open Protocol frame from raw ASCII data.
 * @param {string|Buffer} data
 * @returns {{ raw: string, length: string, mid: string, revision: string, body: string }}
 */
function parseMID(data) {
  const msg = (Buffer.isBuffer(data) ? data.toString("ascii") : data).replace(
    /\x00/g,
    ""
  );

  return {
    raw: msg,
    length: msg.substring(0, 4),
    mid: msg.substring(4, 8),
    revision: msg.substring(8, 11),
    body: msg.substring(HEADER_LENGTH),
  };
}

/**
 * Extract the accepted/rejected MID from a MID 0005 (Command Accepted)
 * or MID 0004 (Command Error) reply.
 * @param {string} raw - Full frame string (NUL-stripped)
 * @returns {string} 4-digit MID that was accepted/rejected
 */
function getReplyMID(raw) {
  return raw.substring(20, 24);
}

/**
 * Parse MID 0061 - Last Tightening Result Data
 * @param {string} data - Raw ASCII frame
 * @returns {object} Parsed tightening result fields
 */
function parseMID0061(data) {
  const clean = (typeof data === "string" ? data : data.toString("ascii")).replace(
    /\x00/g,
    ""
  );

  const safeParse = (val, divisor = 1) => {
    const num = parseInt(val);
    return isNaN(num) ? null : num / divisor;
  };

  return {
    vin: clean.substring(59, 84).trim(),

    pSet: safeParse(clean.substring(90, 93)),

    torqueMin: safeParse(clean.substring(116, 122), 100),
    torqueMax: safeParse(clean.substring(124, 130), 100),
    torqueTarget: safeParse(clean.substring(132, 138), 100),
    torqueValue: safeParse(clean.substring(140, 146), 100),

    angleMin: safeParse(clean.substring(148, 153)),
    angleMax: safeParse(clean.substring(155, 160)),
    angleTarget: safeParse(clean.substring(162, 167)),
    angleValue: safeParse(clean.substring(169, 174)),

    tighteningStatus: parseInt(clean.substring(107, 108)) || 0,
  };
}

module.exports = {
  buildFrame,
  buildMID0001,
  buildMID0050,
  buildMID0060,
  buildMID0062,
  buildMID9999,
  parseMID,
  parseMID0061,
  getReplyMID,
};

File: go/open-protocol/openprotocol.go
// Package openprotocol is a Go port of the Node @digta/open-protocol library:
// an Atlas Copco Open Protocol MID codec plus an ergonomic TCP client
// (auto-reconnect, heartbeat, tightening callbacks). Zero external deps.
package openprotocol

import (
	"bufio"
	"net"
	"strconv"
	"strings"
	"sync"
	"time"
)

// --- codec ---

// BuildFrame builds a raw MID frame: length header + MID + revision + spare, then
// body and a NUL terminator. Mirrors the Node buildFrame (length includes NUL).
func BuildFrame(mid, revision, body string) []byte {
	total := 20 + len(body) + 1
	header := pad(strconv.Itoa(total), 4) + mid + pad(revision, 3) + strings.Repeat(" ", 9)
	return []byte(header + body + "\x00")
}

// The canned MID builders match the exact byte strings the Node library sends.
func BuildMID0001() []byte { return []byte("00200001001" + strings.Repeat(" ", 9) + "\x00") }
func BuildMID0060() []byte { return []byte("002000600011" + strings.Repeat(" ", 8) + "\x00") }
func BuildMID0062() []byte { return []byte("00200062" + strings.Repeat(" ", 12) + "\x00") }
func BuildMID9999() []byte { return []byte("00209999001" + strings.Repeat(" ", 9) + "\x00") }

// BuildMID0050 builds a Vehicle ID (VIN) download; vin is padded/truncated to 25.
func BuildMID0050(vin string) []byte {
	if len(vin) > 25 {
		vin = vin[:25]
	}
	vin += strings.Repeat(" ", 25-len(vin))
	return []byte("00450050" + strings.Repeat(" ", 12) + vin + "\x00")
}

// Frame is a parsed MID frame.
type Frame struct {
	Raw, Length, MID, Revision, Body string
}

// ParseMID parses an incoming frame (NUL stripped).
func ParseMID(data []byte) Frame {
	c := stripNUL(data)
	return Frame{Raw: c, Length: sub(c, 0, 4), MID: sub(c, 4, 8), Revision: sub(c, 8, 11), Body: sub(c, 20, len(c))}
}

// GetReplyMID extracts the replied MID from a MID 0005/0004 response body.
func GetReplyMID(raw string) string { return sub(raw, 20, 24) }

// TighteningResult is the parsed payload of a MID 0061.
type TighteningResult struct {
	VIN                                             string
	PSet                                            int
	TorqueMin, TorqueMax, TorqueTarget, TorqueValue float64
	AngleMin, AngleMax, AngleTarget, AngleValue     int
	TighteningStatus                                int
}

// ParseMID0061 parses a Last Tightening Result frame.
func ParseMID0061(data []byte) TighteningResult {
	c := stripNUL(data)
	return TighteningResult{
		VIN:              strings.TrimSpace(sub(c, 59, 84)),
		PSet:             atoi(sub(c, 90, 93)),
		TorqueMin:        float64(atoi(sub(c, 116, 122))) / 100,
		TorqueMax:        float64(atoi(sub(c, 124, 130))) / 100,
		TorqueTarget:     float64(atoi(sub(c, 132, 138))) / 100,
		TorqueValue:      float64(atoi(sub(c, 140, 146))) / 100,
		AngleMin:         atoi(sub(c, 148, 153)),
		AngleMax:         atoi(sub(c, 155, 160)),
		AngleTarget:      atoi(sub(c, 162, 167)),
		AngleValue:       atoi(sub(c, 169, 174)),
		TighteningStatus: atoi(sub(c, 107, 108)),
	}
}

func stripNUL(b []byte) string { return strings.ReplaceAll(string(b), "\x00", "") }
func sub(s string, start, end int) string {
	if start >= len(s) {
		return ""
	}
	if end > len(s) {
		end = len(s)
	}
	return s[start:end]
}
func atoi(s string) int { n, _ := strconv.Atoi(strings.TrimSpace(s)); return n }
func pad(s string, n int) string {
	for len(s) < n {
		s = "0" + s
	}
	return s
}

// --- client ---

// Client is an ergonomic Open Protocol TCP client. Set the callbacks, then
// call Connect. It runs the MID 0001/0002/0060/0005 handshake, keeps a MID 9999
// heartbeat, auto-acknowledges results, and reconnects on drop.
type Client struct {
	addr              string
	heartbeatInterval time.Duration
	reconnectDelay    time.Duration
	autoAck           bool
	autoSubscribe     bool

	// Callbacks (all optional). Invoked from the client's read goroutine.
	OnConnect    func()
	OnReady      func()
	OnTightening func(TighteningResult)
	OnDisconnect func()
	OnError      func(error)

	mu      sync.Mutex
	conn    net.Conn
	writeMu sync.Mutex
	lastVIN string
	closed  bool
}

// Option configures a Client.
type Option func(*Client)

func WithHeartbeatInterval(d time.Duration) Option {
	return func(c *Client) { c.heartbeatInterval = d }
}
func WithReconnectDelay(d time.Duration) Option { return func(c *Client) { c.reconnectDelay = d } }
func WithAutoAck(v bool) Option                 { return func(c *Client) { c.autoAck = v } }
func WithAutoSubscribe(v bool) Option           { return func(c *Client) { c.autoSubscribe = v } }

// New creates a client for addr ("host:port").
func New(addr string, opts ...Option) *Client {
	c := &Client{
		addr:              addr,
		heartbeatInterval: 10 * time.Second,
		reconnectDelay:    time.Second,
		autoAck:           true,
		autoSubscribe:     true,
	}
	for _, o := range opts {
		o(c)
	}
	return c
}

// Connect starts the client (non-blocking); it dials and keeps reconnecting
// until Close is called.
func (c *Client) Connect() {
	go c.loop()
}

// SendVin sends a VIN (MID 0050); it is stored and re-sent after reconnect.
func (c *Client) SendVin(vin string) {
	c.mu.Lock()
	c.lastVIN = vin
	c.mu.Unlock()
	c.send(BuildMID0050(vin))
}

// Close stops the client and its reconnect loop.
func (c *Client) Close() {
	c.mu.Lock()
	c.closed = true
	conn := c.conn
	c.mu.Unlock()
	if conn != nil {
		conn.Close()
	}
}

func (c *Client) loop() {
	for {
		if c.isClosed() {
			return
		}
		err := c.session()
		if c.OnDisconnect != nil {
			c.OnDisconnect()
		}
		if err != nil && !c.isClosed() && c.OnError != nil {
			c.OnError(err)
		}
		if c.isClosed() {
			return
		}
		time.Sleep(c.reconnectDelay)
	}
}

func (c *Client) session() error {
	conn, err := net.DialTimeout("tcp", c.addr, 5*time.Second)
	if err != nil {
		return err
	}
	c.mu.Lock()
	c.conn = conn
	c.mu.Unlock()
	defer conn.Close()

	if c.OnConnect != nil {
		c.OnConnect()
	}
	if err := c.write(conn, BuildMID0001()); err != nil {
		return err
	}

	hbStop := make(chan struct{})
	var hbOnce sync.Once
	stopHB := func() { hbOnce.Do(func() { close(hbStop) }) }
	defer stopHB()

	r := bufio.NewReader(conn)
	for {
		frame, err := r.ReadBytes(0x00) // Open Protocol frames are NUL-terminated
		if err != nil {
			return err
		}
		f := ParseMID(frame)
		switch f.MID {
		case "0002": // comm start acknowledged -> subscribe
			if c.autoSubscribe {
				c.write(conn, BuildMID0060())
			}
		case "0005": // command accepted
			if GetReplyMID(f.Raw) == "0060" {
				c.startHeartbeat(conn, hbStop)
				if c.OnReady != nil {
					c.OnReady()
				}
				if vin := c.getVIN(); vin != "" {
					c.write(conn, BuildMID0050(vin))
				}
			}
		case "0061": // tightening result
			if c.OnTightening != nil {
				c.OnTightening(ParseMID0061(frame))
			}
			if c.autoAck {
				c.write(conn, BuildMID0062())
			}
		}
	}
}

func (c *Client) startHeartbeat(conn net.Conn, stop <-chan struct{}) {
	go func() {
		t := time.NewTicker(c.heartbeatInterval)
		defer t.Stop()
		for {
			select {
			case <-stop:
				return
			case <-t.C:
				if err := c.write(conn, BuildMID9999()); err != nil {
					return
				}
			}
		}
	}()
}

// write serializes writes to a specific connection.
func (c *Client) write(conn net.Conn, frame []byte) error {
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	conn.SetWriteDeadline(time.Now().Add(2 * time.Second))
	_, err := conn.Write(frame)
	return err
}

// send writes to the current connection if any.
func (c *Client) send(frame []byte) {
	c.mu.Lock()
	conn := c.conn
	c.mu.Unlock()
	if conn != nil {
		c.write(conn, frame)
	}
}

func (c *Client) isClosed() bool { c.mu.Lock(); defer c.mu.Unlock(); return c.closed }
func (c *Client) getVIN() string { c.mu.Lock(); defer c.mu.Unlock(); return c.lastVIN }
